stdlib/public/Darwin/Foundation/Data.swift

```diff
@@ -1271,11 +1272,14 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
             storage.copyBytes(to: pointer, from: range)
         }
 
-        @inlinable
-        var hashValue: Int {
-            let hashRange = startIndex..<Swift.min(startIndex + 80, endIndex)
-            return storage.withUnsafeBytes(in: hashRange) {
-                return Int(bitPattern: CFHashBytes(UnsafeMutablePointer(mutating: $0.baseAddress?.assumingMemoryBound(to: UInt8.self)), CFIndex($0.count)))
+        @inlinable @inline(__always) // This should always be inlined into _Representation.hash(into:).
+        func hash(into hasher: inout Hasher) {
+            hasher.combine(count)
+
+            // Hash at most the first 80 bytes of this data.
+            let range = startIndex ..< Swift.min(startIndex + 80, endIndex)
+            storage.withUnsafeBytes(in: range) {
+                hasher.combine(bytes: $0)
```

Comment:
The above comment applies here, too -- all bytes should be hashed.

Reply:
To reiterate what Philippe mentions above — hashing need not be O(N) to be effective, and we often do this for collections to avoid O(N) hashing. This matches what we do for `NSData`, `NSString`, our collections, etc.

Reply:
The only way to guarantee effective hashing is to feed the hasher with all information that is compared in `==`. It is inappropriate for `Data` to silently assume that the first 80 bytes (or any such prefix) will distinguish between its instances.

This also applies to Foundation classes, and it applies particularly strongly to collection types like `NSSet` (cf. https://github.com/apple/swift/pull/15382). I do accept that changing how those are hashed may not be feasible -- but it would be highly unfortunate if we emulated these implementations in newer types.